Mathlib/Geometry/Manifold/Notation.lean

```diff
@@ -0,0 +1,471 @@
+            if ← withReducible (pureIsDefEq E F) then return some K else return none
+          | _ => return none
+        | throwError "Couldn't find a `NormedSpace` structure on {F} among local instances."
+      trace[Elab.DiffGeo.MDiff] "{F} is a normed field over {K}"
+      let kT : Term ← Term.exprToSyntax K
+      let srcIT : Term ← Term.exprToSyntax srcI
+      let FT : Term ← Term.exprToSyntax F
+      let iTerm : Term ← ``(ModelWithCorners.prod $srcIT 𝓘($kT, $FT))
+      Term.elabTerm iTerm none
+    else
+      throwError "No `baseInfo` provided"
+  fromNormedSpace : TermElabM Expr := do
+    let some (inst, K) ← findSomeLocalInstanceOf? ``NormedSpace fun inst type ↦ do
+        match_expr type with
+        | NormedSpace K E _ _ =>
+          if ← withReducible (pureIsDefEq E e) then return some (inst, K) else return none
+        | _ => return none
+      | throwError "Couldn't find a `NormedSpace` structure on {e} among local instances."
+    trace[Elab.DiffGeo.MDiff] "Field is: {K}"
+    mkAppOptM ``modelWithCornersSelf #[K, none, e, none, inst]
+  fromChartedSpace : TermElabM Expr := do
+    let some H ← findSomeLocalInstanceOf? ``ChartedSpace fun _ type ↦ do
+        match_expr type with
+        | ChartedSpace H _ M _ =>
+          if ← withReducible (pureIsDefEq M e) then return some H else return none
+        | _ => return none
+      | throwError "Couldn't find a `ChartedSpace` structure on {e} among local instances."
+    trace[Elab.DiffGeo.MDiff] "H is: {H}"
+    let some m ← findSomeLocalHyp? fun fvar type ↦ do
+        match_expr type with
+        | ModelWithCorners _ _ _ _ _ H' _ => do
+          if ← withReducible (pureIsDefEq H' H) then return some fvar else return none
+        | _ => return none
+      | throwError "Couldn't find a `ModelWithCorners` with model space {H} in the local context."
+    return m
+  fromNormedField : TermElabM Expr := do
+    let eT : Term ← Term.exprToSyntax e
+    let iTerm : Term ← ``(𝓘($eT, $eT))
+    Term.elabTerm iTerm none
+
+/-- If the type of `e` is a non-dependent function between spaces `src` and `tgt`, try to find a
+model with corners on both `src` and `tgt`. If successful, return both models.
+
+We pass `e` instead of just its type for better diagnostics.
+
+If `es` is `some`, we verify that `src` and the type of `es` are definitionally equal. -/
+def findModels (e : Expr) (es : Option Expr) : TermElabM (Expr × Expr) := do
+  let etype ← whnf <|← instantiateMVars <|← inferType e
+  match etype with
+  | .forallE _ src tgt _ =>
+    if tgt.hasLooseBVars then
+      -- TODO: try `T%` here, and if it works, add an interactive suggestion to use it
+      throwError "Term {e} is a dependent function, of type {etype}\nHint: you can use the `T%` \
+        elaborator to convert a dependent function to a non-dependent one"
+    let srcI ← findModel src
+    if let some es := es then
+      let estype ← inferType es
+      if !(← pureIsDefEq estype <|← mkAppM ``Set #[src]) then
```

Comment:
I reflected on my suggestion to use `pureIsDefEq` here, and changed my mind: I think `isDefEq` is appropriate. We might actually even like to be more aggressive; I've left that as a TODO.

```suggestion
      /- Note: we use `isDefEq` here since persistent metavariable assignments in `src` and 
      `estype` are acceptable.
      TODO: consider attempting to coerce `es` to a `Set`. -/
      if !(← isDefEq estype <|← mkAppM ``Set #[src]) then
```